src/Dapr.AspNetCore/DaprEndpointRouteBuilderExtensions.cs

```diff
@@ -33,14 +36,27 @@ public static IEndpointConventionBuilder MapSubscribeHandler(this IEndpointRoute
             {
                 var dataSource = context.RequestServices.GetRequiredService<EndpointDataSource>();
                 var entries = dataSource.Endpoints
-                    .Select(e => e.Metadata.GetMetadata<TopicAttribute>()?.Name)
-                    .Where(n => n != null)
+                    .OfType<RouteEndpoint>()
+                    .Where(e => e.Metadata.GetMetadata<TopicAttribute>()?.Name != null)   // only endpoints which have  TopicAttribute with not null Name.
+                    .Where(e => e.RoutePattern.Parameters.Count == 0) // only endpoints which don't have parameters.
                     .Distinct()
-                    .ToArray();
+                    .Select(e => (e.Metadata.GetMetadata<TopicAttribute>().Name, e.RoutePattern.RawText));
 
                 context.Response.ContentType = "application/json";
-                await JsonSerializer.SerializeAsync(context.Response.Body, entries, context.RequestServices.GetService<JsonSerializerOptions>());
+                using Utf8JsonWriter writer = new Utf8JsonWriter(context.Response.BodyWriter);
+                writer.WriteStartArray();
+
+                foreach (var entry in entries)
+                {
+                    writer.WriteStartObject();
+                    writer.WriteString("topic", entry.Name);
+                    writer.WriteString("route", entry.RawText);
```

Comment:
Does `route` expect a leading slash? does it tolerate one? (same for trailing). 

`RawText` will be whatever the developer typed in. You might also need to deal with the case where someone wrote `~/foo/` - which is a legacy syntax routing accepts. 

For a non-parameterized route you might want to normalize it with something like:

```C#
string.Join("/", pattern.PathSegments.Select(segment => string.Concat(segment.Parts.Cast<RoutePatternLiteralPart>().Select(part => part.Content))));
```

This will combine all of the segments using `/` as a separator but without a leading or trailing slash.

Reply:
Fixing it.